actors/evm/src/interpreter/precompiles.rs

```diff
@@ -93,7 +103,75 @@ fn read_right_pad<'a>(input: impl Into<Cow<'a, [u8]>>, len: usize) -> Cow<'a, [u
     input
 }
 
-// https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/core/vm/contracts.go#L165
+fn todo<RT: Runtime>(_: &RT, _: &[u8]) -> PrecompileResult {
+    todo!()
+}
+
+fn get_actor_code_cid<RT: Runtime>(rt: &RT, input: &[u8]) -> PrecompileResult {
+    let id = EthAddress(input[..20].try_into().unwrap()).as_id();
+    if id.is_none() {
+        return Err(PrecompileError::InvalidInput);
+    }
+    Ok(rt.get_actor_code_cid(&id.unwrap()).unwrap_or_default().to_bytes())
+}
+
+fn get_randomness<RT: Runtime>(rt: &RT, input: &[u8]) -> PrecompileResult {
+    // 1 + 7 (reserved) + 8 + 8 + 8 = 32
+    let word = read_right_pad(input, 32);
+
+    #[derive(num_derive::FromPrimitive)]
+    #[repr(u8)]
+    enum RandomnessType {
+        Chain = 0,
+        Beacon = 1,
+    }
+
+    let randomness_type = RandomnessType::from_u8(word[0]);
+    // 7 bytes reserved
+    let personalization = i64::from_le_bytes(word[8..16].try_into().unwrap());
+    let rand_epoch = i64::from_le_bytes(word[16..24].try_into().unwrap());
+    let entropy_len = u64::from_le_bytes(word[24..32].try_into().unwrap());
+
+    let entropy = read_right_pad(&input[32..], entropy_len as usize);
```

Comment:
Any reason to specify a length instead of just reading the rest?

Reply:
zero padding things... we can say we don't follow EVM convention and whatever bytes you put in is whatever bytes that get used. I am un-opinionated here...

Reply:
The EVM convention applies to:

- Fixed sized values.
- Multiple variable length values.

For example:

- Identity doesn't do this.
- The hash functions don't do this.

Reply:
On the other hand... we could just say "well, we have multiple parameters."

Reply:
Ok, fine, I'm convinced. We have multiple parameters, let's follow that convention.